File: smartjs/flashnchips/slip.cc
#include "slip.h"

#include <QDebug>

#include "status_qt.h"

namespace SLIP {

// https://tools.ietf.org/html/rfc1055
const unsigned char SLIPFrameDelimiter = 0xC0;
const unsigned char SLIPEscape = 0xDB;
const unsigned char SLIPEscapeFrameDelimiter = 0xDC;
const unsigned char SLIPEscapeEscape = 0xDD;

util::Status send(QSerialPort *port, const QByteArray &data, int timeoutMs) {
  const QString prefix = QString("SLIP::send(%1, %2, %3):")
                             .arg(port->portName())
                             .arg(data.length())
                             .arg(timeoutMs);
  qDebug() << prefix << "=>" << data.toHex();
  bool ok = port->putChar(SLIPFrameDelimiter);
  for (int i = 0; ok && i < data.length(); i++) {
    switch ((unsigned char) data[i]) {
      case SLIPFrameDelimiter:
        ok = ok && port->putChar(SLIPEscape);
        ok = ok && port->putChar(SLIPEscapeFrameDelimiter);
        break;
      case SLIPEscape:
        ok = ok && port->putChar(SLIPEscape);
        ok = ok && port->putChar(SLIPEscapeEscape);
        break;
      default:
        ok = ok && port->putChar(data[i]);
        break;
    }
  }
  ok = ok && port->putChar(SLIPFrameDelimiter);
  ok = ok && port->waitForBytesWritten(timeoutMs);
  if (!ok) {
    return QS(util::error::UNAVAILABLE, prefix + " " + port->errorString());
  }
  return util::Status::OK;
}

util::StatusOr<QByteArray> recv(QSerialPort *port, int timeoutMs) {
  const QString prefix =
      QString("SLIP::recv(%1, %2): ").arg(port->portName()).arg(timeoutMs);
  QByteArray ret;
  char c = 0;
  // Skip everything before the frame start.
  do {
    if (port->bytesAvailable() == 0 && !port->waitForReadyRead(timeoutMs)) {
      return QS(util::error::UNAVAILABLE,
                prefix + "no data 1: " + port->errorString());
    }
    if (!port->getChar(&c)) {
      return QS(util::error::UNAVAILABLE,
                prefix + "failed to read prefix: " + port->errorString());
    }
  } while ((unsigned char) c != SLIPFrameDelimiter);
  for (;;) {
    if (port->bytesAvailable() == 0 && !port->waitForReadyRead(timeoutMs)) {
      return QS(util::error::UNAVAILABLE,
                prefix + "no data 2: " + port->errorString());
    }
    if (!port->getChar(&c)) {
      return QS(util::error::UNAVAILABLE, prefix + "failed to read next char");
    }
    switch ((unsigned char) c) {
      case SLIPFrameDelimiter:
        // End of frame.
        qDebug() << "<=" << ret.toHex();
        return ret;
      case SLIPEscape:
        if (port->bytesAvailable() == 0 && !port->waitForReadyRead(timeoutMs)) {
          return QS(util::error::UNAVAILABLE,
                    prefix + "no data 3: " + port->errorString());
        }
        if (!port->getChar(&c)) {
          return QS(util::error::UNAVAILABLE,
                    prefix + "failed to read next char");
        }
        switch ((unsigned char) c) {
          case SLIPEscapeFrameDelimiter:
            ret.append(SLIPFrameDelimiter);
            break;
          case SLIPEscapeEscape:
            ret.append(SLIPEscape);
            break;
          default:
            return QS(util::error::UNAVAILABLE,
                      prefix + "invalid escape sequence: " + int(c));
        }
        break;
      default:
        ret.append(c);
        break;
    }
  }
}

}  // namespace SLIP

File: mos/flash/common/slip.go
package common

import (
	"io"

	"github.com/cesanta/errors"
	"github.com/golang/glog"
)

const (
	// https://tools.ietf.org/html/rfc1055
	slipFrameDelimiter       = 0xC0
	slipEscape               = 0xDB
	slipEscapeFrameDelimiter = 0xDC
	slipEscapeEscape         = 0xDD
)

type SLIPReaderWriter struct {
	rw io.ReadWriter
}

func NewSLIPReaderWriter(rw io.ReadWriter) *SLIPReaderWriter {
	return &SLIPReaderWriter{rw: rw}
}

func (srw *SLIPReaderWriter) Read(buf []byte) (int, error) {
	n := 0
	start := true
	esc := false
	for {
		b := []byte{0}
		bn, err := srw.rw.Read(b)
		if err != nil || bn != 1 {
			return n, errors.Annotatef(err, "error reading")
		}
		if start {
			if b[0] != slipFrameDelimiter {
				return 0, errors.Errorf("invalid SLIP starting byte: 0x%02x", b[0])
			}
			start = false
			continue
		}
		if !esc {
			switch b[0] {
			case slipFrameDelimiter:
				glog.V(4).Infof("<= (%d) %s", n, LimitStr(buf[:n], 32))
				return n, nil
			case slipEscape:
				esc = true
			default:
				if n >= len(buf) {
					return n, errors.Errorf("frame buffer overflow (%d)", len(buf))
				}
				buf[n] = b[0]
				n += 1
			}
		} else {
			if n >= len(buf) {
				return n, errors.Errorf("frame buffer overflow (%d)", len(buf))
			}
			switch b[0] {
			case slipEscapeFrameDelimiter:
				buf[n] = slipFrameDelimiter
			case slipEscapeEscape:
				buf[n] = slipEscape
			default:
				return n, errors.Errorf("invalid SLIP escape sequence: %d", b[0])
			}
			n += 1
			esc = false
		}
	}
}

func (srw *SLIPReaderWriter) Write(data []byte) (int, error) {
	frame := []byte{slipFrameDelimiter}
	for _, b := range data {
		switch b {
		case slipFrameDelimiter:
			frame = append(frame, slipEscape)
			frame = append(frame, slipEscapeFrameDelimiter)
		case slipEscape:
			frame = append(frame, slipEscape)
			frame = append(frame, slipEscapeEscape)
		default:
			frame = append(frame, b)
		}
	}
	frame = append(frame, slipFrameDelimiter)
	glog.V(4).Infof("=> (%d) %s", len(data), LimitStr(data, 32))
	return srw.rw.Write(frame)
}
